src/core/symbology-ng/qgsfillsymbollayerv2.cpp

```diff
@@ -3432,9 +3436,38 @@ void QgsCentroidFillSymbolLayerV2::renderPolygon( const QPolygonF& points, QList
 {
   Q_UNUSED( rings );
 
-  if ( !mPointOnAllParts &&
-       context.renderContext().expressionContext().variable( QgsExpressionContext::EXPR_GEOMETRY_PART_NUM ).toInt() != context.renderContext().expressionContext().variable( QgsExpressionContext::EXPR_GEOMETRY_PART_BIGGEST ).toInt() )
-    return;
+  if ( !mPointOnAllParts )
+  {
+    const QgsFeature* feature = context.feature();
+    if ( feature )
+    {
+      if ( feature->id() != mBiggestPartFeatureId )
+      {
+        mBiggestPartFeatureId = feature->id();
+        mBiggestPartIndex = 1;
+
+        unsigned int part_count = context.renderContext().expressionContext().variable( QgsExpressionContext::EXPR_GEOMETRY_PART_COUNT ).toUInt();
+        if ( part_count > 1 )
+        {
+          const QgsGeometry *geom = feature->constGeometry();
+          const QgsGeometryCollectionV2* geomCollection = dynamic_cast<const QgsGeometryCollectionV2*>( geom->geometry() );
+
+          double area_biggest = 0;
+          for ( unsigned int i = 0; i < part_count; ++i )
+          {
+            if ( geomCollection->geometryN( i )->area() > area_biggest )
+            {
+              area_biggest = geomCollection->geometryN( i )->area();
+              mBiggestPartIndex = i + 1;
+            }
+          }
+        }
+      }
+
+      if ( context.renderContext().expressionContext().variable( QgsExpressionContext::EXPR_GEOMETRY_PART_NUM ).toInt() != mBiggestPartIndex )
```

Comment:
That's the intention - set it once for expressions in the expression context, and using the symbol context for non-expression usage. I'm basically wanting to avoid expression contexts being used as a shortcut way to pass values around like this.